connectors/uplynk/src/main/java/com/theoplayer/android/connector/uplynk/internal/UplynkSsaiDescriptionConverter.kt

```diff
@@ -1,31 +1,78 @@
 package com.theoplayer.android.connector.uplynk.internal
 
+import com.theoplayer.android.connector.uplynk.UplynkAssetType
 import com.theoplayer.android.connector.uplynk.UplynkSsaiDescription
+import kotlin.time.Duration
+
+private const val AD_IMPRESSIONS = 1
+private const val FW_VIDEO_VIEWS = 2
+private const val LINEAR_AD_DATA = 4
 
 internal class UplynkSsaiDescriptionConverter {
     private val DEFAULT_PREFIX = "https://content.uplynk.com"
 
-    fun buildPreplayUrl(ssaiDescription: UplynkSsaiDescription): String = with(ssaiDescription) {
+    fun buildPreplayVodUrl(ssaiDescription: UplynkSsaiDescription): String = with(ssaiDescription) {
         val prefix = prefix ?: DEFAULT_PREFIX
-        val assetIds = when {
-            assetIds.isEmpty() && externalId.size == 1 -> "$userId/${externalId.first()}.json"
-            assetIds.isEmpty() && externalId.size > 1 -> "$userId/${externalId.joinToString(",")}/multiple.json"
-            assetIds.size == 1 -> "${assetIds.first()}.json"
-            else -> assetIds.joinToString(separator = ",") + "/multiple.json"
+
+        var url = "$prefix/preplay/$urlAssetId?v=2"
+        if (ssaiDescription.contentProtected) {
+            url += "&manifest=mpd"
+            url += "&rmt=wv"
         }
 
-        var url = "$prefix/preplay/$assetIds?v=2"
+        url += "&$pingParameters&$urlParameters"
+
+        return url
+    }
+
+    fun buildPreplayLiveUrl(ssaiDescription: UplynkSsaiDescription): String = with(ssaiDescription) {
+        val prefix = prefix ?: DEFAULT_PREFIX
+
+        var url = "$prefix/preplay/$urlAssetType/$urlAssetId?v=2"
         if (ssaiDescription.contentProtected) {
             url += "&manifest=mpd"
             url += "&rmt=wv"
         }
 
-        val parameters = preplayParameters.map { "${it.key}=${it.value}" }.joinToString("&")
-        url += "&$parameters"
+        url += "&$pingParameters&$urlParameters"
 
         return url
     }
 
+    private val UplynkSsaiDescription.urlParameters
+        get() = preplayParameters.map { "${it.key}=${it.value}" }.joinToString("&")
+
+    private val UplynkSsaiDescription.pingParameters: String
+        get() {
+            val isLive = assetType == UplynkAssetType.ASSET
+
+            val features = with(pingConfiguration) {
+                (AD_IMPRESSIONS.takeIf { !isLive && adImpressions }
+                    ?: 0) + (FW_VIDEO_VIEWS.takeIf { !isLive && freeWheelVideoViews }
+                    ?: 0) + (LINEAR_AD_DATA.takeIf { isLive && linearAdData } ?: 0)
+            }
```

Comment:
Nitpick: plz make it more readable

Reply:
Check UplynkPingFeatures. This logic is there now